components/encryption/src/master_key/kms.rs

```diff
@@ -99,37 +105,46 @@ impl AwsKms {
             client
                 .generate_data_key(generate_request.clone())
                 .map_err(|e| Error::Other(e.into()))
-        });
+        })
+        .unwrap();
         let ciphertext_key = generate_response.ciphertext_blob.unwrap().as_ref().to_vec();
         let plaintext_key = generate_response.plaintext.unwrap().as_ref().to_vec();
         Ok((ciphertext_key, plaintext_key))
     }
 }
 
-fn retry<T, U, F>(runtime: &mut Runtime, mut func: F) -> T
+fn retry<T, U, F>(runtime: &mut Runtime, mut func: F) -> Result<T>
 where
     F: FnMut() -> U,
     U: Future<Output = Result<T>> + std::marker::Unpin,
 {
     let retry_limit = 6;
     let timeout_duration = Duration::from_secs(10);
+    let mut last_err = None;
     for _ in 0..retry_limit {
         let fut = func();
 
         match runtime.block_on(async move {
             let timeout = tokio::time::delay_for(timeout_duration);
             future::select(fut, timeout).await
         }) {
-            future::Either::Left((Ok(resp), _)) => return resp,
+            future::Either::Left((Ok(resp), _)) => return Ok(resp),
             future::Either::Left((Err(e), _)) => {
                 error!("kms request failed"; "error"=>?e);
+                if let Error::WrongMasterKey(e) = e {
+                    return Err(Error::WrongMasterKey(e));
+                }
+                last_err = Some(e);
             }
             future::Either::Right((_, _)) => {
                 error!("kms request timeout"; "timeout" => ?timeout_duration);
             }
         }
     }
-    panic!("kms request failed in {} times", retry_limit)
+    panic!(
```

Comment:
why panic? how about just return error?

Reply:
How to handle the error?

Reply:
`retry` do return result. Just return Error::Other with proper message?